apex/contrib/groupbn/batch_norm.py

```diff
@@ -159,21 +196,27 @@ def forward(self, x, z=None):
             return bn_addrelu_NHWC_impl.apply(x, z,
                                   self.weight, self.bias,
                                   self.running_mean, self.running_var,
-                                  self.minibatch_mean, self.minibatch_riv,
+                                  self.minibatch_mean, self.minibatch_riv, self.grid_dim_y, self.ret_cta,
                                   self.momentum,
-                                  self.eps, self.training, self.bn_group, self.my_data, self.pair_data, (self.magic), self.pair_data2,
-                                  self.max_cta_per_sm, self.cta_launch_margin)
+                                  self.eps, self.training, self.bn_group, self.my_data, self.pair_data, (self.magic), self.pair_data2, self.pair_data3,
+                                  self.addrelu_fwd_occupancy, self.addrelu_fwd_grid_dim_x,
```

Comment:
Is there a reason for needed these (occupancy, grid dim) to be exposed & stored in python-land? The implementations look cheap enough to do each time and it'd clean up the interface (along with not needing to expose them through pybind)

Reply:
Occupancy and grid_dim_x are determined by querying device properties. Moving these to pytorch allowed reducing to a single device query in construction time instead of on every layer execution. Given it is theoretically possible one will use a single process driving more than one GPU, a simple static variable caching device properties will not be a sufficiently general solution. As such I preferred moving these to pytorch rather than implementing more complex caching scheme.

Reply:
:+1: